phac-nml/irida-next
# Graphql attach files to sample mutation

## What does this PR do and why?

Adds AttachFilesToSample mutation to the GraphQL API
Fixes #391 

## Screenshots or screen recordings

N/A

## How to set up and validate locally

1. `./scripts/clean_dev.sh`
2. Login as `user1@email.com`
3. Navigate to `http://localhost:3000/graphiql` to open the graphql interface .
4. Verify you can query sample from graphql and rails console

This graphql query gets you sample 51
```graphql
{
query getonesamp {
  node(id: "gid://irida/Sample/51"){
    ... on Sample{
      id,
      name,
      description,
      metadata
    }
  }
}
}
```

In a terminal, open rails console with `rails c`

The following command gets you the sample object matching above
```ruby
> Sample.all[59]
```
querying for attachments should result in an empty list
```ruby
> Sample.all[59].attachments
```

5. find a file to upload, get the content type, md5 checksum and bytesize. Or download this file and use the provided values in step 6

[myfile.json](https://github.com/phac-nml/irida-next/files/14129150/myfile.json)

6. Create a blob for your file

```graphql
mutation makeblob {
  createDirectUpload(
    input: {
      filename: "myfile.json", # file name
      contentType: "application/json", # file content type
      checksum: "Y33CgI35hFoI6p+WBXYl+A==", # md5
      byteSize: 151 # size in bytes
    }
  ) {
    directUpload {
      url,
      headers,
      blobId,
      signedBlobId
    }
  }
}
```

We will use the `url`, `headers` and `signedBlobId`

7. Upload your file. This is simplest with Postman
Make a `PUT` request, putting the returned `url` in the URL field

![image](https://github.com/phac-nml/irida-next/assets/16961365/d960b0c3-706e-4311-9986-e512ab7731a5)

Attach the file as binary

![image](https://github.com/phac-nml/irida-next/assets/16961365/b3d0f8af-269b-4b71-820d-bdf8f5d99db7)

Verify your content type header matches the `headers` from above. Postman should auto-generate this based on the file you attached.

When `Send` is clicked, you should receive a `204 No Content`

8. Execute the mutation to attach files to a Sample.

```graphql
mutation attachfilestosample {
  attachFilesToSample (
    input: {
      files: ["YOUR signedBlobId FROM ABOVE"],
      sampleId: "gid://irida/Sample/51"
    }
  ) {
    sample{id},
    status,
    errors
  }
}
```
response
```json
{
  "data": {
    "attachFilesToSample": {
      "sample": {
        "id": "gid://irida/Sample/51"
      },
      "status": {
        "eyJfcmFpbHMiOnsiZGF0YSI6NTIsInB1ciI6ImJsb2JfaWQifX0=--1aae6b4a9798a4593e24499df749f20ad862af00": "success"
      },
      "errors": []
    }
  }
}
```

9. Verify that you see all the files you attached

Querying for attachments should result in one attachment, which you uploaded.
```ruby
> Sample.all[59].attachments
```

```ruby
[#<Attachment:0x00007fe0db6ae9d8
  id: 51,
  metadata: {"format"=>"unknown", "compression"=>"none"},
  deleted_at: nil,
  attachable_type: "Sample",
  attachable_id: 51,
  created_at: Thu, 01 Feb 2024 17:59:39.890147000 UTC +00:00,
  updated_at: Thu, 01 Feb 2024 17:59:39.921974000 UTC +00:00,
  log_data: nil>]
```

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
